R/register_handlers.R

```diff
@@ -0,0 +1,125 @@
+#'
+#' @param namespace (`character(1)`) the logger namespace
+#' @param package (`character(1)`) the package name
+#'
+#' @return `NULL` invisibly. Called for its side effects.
+#'
+#' @details
+#' This function registers global handlers for messages, warnings and errors.
+#' The handlers will investigate the call stack and if it contains a function
+#' from the package, the message, warning or error will be logged into the respective
+#' namespace.
+#'
+#' The handlers are registered only once per package and type. Consecutive calls will be ignored.
+#'
+#' Use `TEAL.LOG_MUFFLE` environmental variable or `teal.log_muffle` R option to optionally
+#' control recover strategies. If `TRUE` (a default value) then the handler will jump to muffle
+#' restart for a given type of condition and doesn't continue (with output to the console).
+#' Applicable for message and warning types only and the errors won't be suppressed.
+#'
+#' @note Registering handlers is forbidden within `tryCatch()` or `withCallingHandlers()`.
+#' Because of this, handlers are registered only if it is possible.
+#'
+#' @seealso [globalCallingHandlers()]
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' register_handlers("teal.logger")
+#' # see the outcome
+#' globalCallingHandlers()
+#' }
+register_handlers <- function(namespace, package = namespace) {
+  if (register_handlers_possible()) {
+    register_handler_type(namespace = namespace, package = package, type = "message")
+    register_handler_type(namespace = namespace, package = package, type = "warning")
+    register_handler_type(namespace = namespace, package = package, type = "error")
+  }
+
+  invisible(NULL)
+}
+
+register_handler_type <- function(
+    namespace,
+    package = namespace,
+    type = c("error", "warning", "message")) {
+  if (!(is.character(namespace) && length(namespace) == 1)) {
+    stop("namespace argument must be a single string.")
+  }
+  if (!(namespace %in% logger::log_namespaces())) {
+    stop("namespace argument must be a pre-registered logger namespace.")
+  }
+  if (!(is.character(package) && length(package) == 1)) {
+    stop("package argument must be a single string.")
+  }
+  match.arg(type)
+
+  registered_handlers_namespaces[[package]] <- namespace
+
+  # avoid re-registering handlers
+  gch <- globalCallingHandlers()[names(globalCallingHandlers()) == type]
+  if (length(gch) > 0 && any(sapply(gch, attr, "type") == "teal.logger_handler")) {
+    return(invisible(NULL))
+  }
+
+  # create a handler object
+  # loop through the call stack starting from the bottom (the last call)
+  # if a function is from pre-registered package then log using pre-specified namespace
+  logger_fun <- switch(type,
+    error = logger::log_error,
+    warning = logger::log_warn,
+    message = logger::log_info
+  )
+  handler_fun <- function(m) {
+    i <- sys.nframe() - 1L
+    while (i > 0L) {
+      pkg_sys_fun_i <- methods::getPackageName(environment(sys.function(i)))
```

Comment:
I've noticed a problem when using `load_all(...)`, when loading a package `handler_fun` is invoked and fails when searching for package name of a primitive function.

see for example 
```r
methods::getPackageName(environment(.Primitive("on.exit")))
# Error in as.environment(where) : using 'as.environment(NULL)' is defunct
```

```suggestion
      env_sys_fun_i <- environment(sys.function(i))
      pkg_sys_fun_i <- if (!is.null(env_sys_fun_i)) { # primitive functions don't have environment
        methods::getPackageName(env_sys_fun_i)
      } else {
        ""
      }
```

Reply:
Thanks for the spot and good suggestion! Committing this!